Continue bootstrapmethods parsing after unsupported entry

mgudemann opened the pull request:
Before it was possible to keep some `BootstrapMethods` entries unparsed, after a
currently unsupported was found. This lead to the situation that JBMC ended up
with a wrong file offset which caused problems in the case of class file
attributes being present after the `BootstrapMethods` attribute.

This was not observed using the Oracle or OpenJDK compilers, but currently only
on the Eclipse compiler for Java.

Commit:
Continue parsing after unsupported  BootstrapMethods entry

Before it was possible to keep some `BootstrapMethods` entries unparsed, after a
currently unsupported was found. This lead to the situation that JBMC ended up
with a wrong file offset which caused problems in the case of class file
attributes being present after the `BootstrapMethods` attribute.

This was not observed using the Oracle or OpenJDK compilers, but currently only
on the Eclipse compiler for Java.

Commit:
Add regression test compiled with ecj

thk123 commented on src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1803,7 +1803,7 @@ java_bytecode_parsert::parse_method_handle(const method_handle_infot &entry)
 {
   u2 num_bootstrap_methods = read_u2();
-  for(size_t i = 0; i < num_bootstrap_methods; i++)
+  for(size_t j = 0; j < num_bootstrap_methods; j++)
```
Why has this changed from `i` to `j`

thk123 replied:
Because it was colliding with another loop, perhaps a good argument for a better name like `boostrap_method_index`?

thk123 commented on src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1868,72 +1868,81 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
+        const pool_entryt &method_handle_argument = pool_entry(argument_index2);
+        const pool_entryt &method_type_argument = pool_entry(argument_index3);
+
+        if(
```
It is a bit hard to follow these nested if/else statements, consider changing to:
```c++
if(! /* validity check 1 */) 
{
  report_error1();
  continue;
}
if(! /* validity check 2 */) 
{
  report_error2();
  continue;
}
// do the real work!
```

majakusber replied:
+1 Some commentary on the individual cases would make it easier to follow too.

thk123 commented on regression/cbmc-java/lambda2/test_no_crash_ecj.desc:
```diff
@@ -0,0 +1,9 @@
+^SIGNAL=0
+--
+--
+Tests that it doesn't crash when failing to read a `BootstrapMethods` entry and
```
Is this compiled with eclipse compiler? If so probably worth making a note of that here to stop people inadvertently updating it.

thk123 replied:
It appears to be - will update docs

majakusber commented on regression/cbmc-java/lambda2/test_no_crash_ecj.desc:
```diff
@@ -0,0 +1,9 @@
+CORE
+StaticMethodRef.class
+--function StaticMethodRef.test2
```
Why this method is tested instead of `Smr`? Wouldn't `Smr` do the same job, with lambdas being ignored but parsed?

thk123 replied:
I agree and the test passes setting to the the `Smr` so will change to that but @mgudemann when you're back if you could comment on whether this was the right thing to do.

majakusber reviewed (changes requested):
The if-else statements are nested now which makes them hard to read. It would be useful for future work on lambdas to make it more readable.

Commit:
Rename j to boostrap_method_index

Commit:
Replace nested else ifs with a continue on error conditions

Commit:
Refactored error reporting method into a seperate function

Commit:
Removed impossible condition

Either an optional was returned or a correct lamdba so no need to handle
the third case. Added an invariant to catch the case this changes.

Commit:
Renamed parameter variables

Commit:
Moved the first condition to after the explanation of our assumptions

Here it is clearer why we need at least 3 arguments. Also documented
this explictly above the condition. Also changed to a debug as there is
nothing in the java spec requiring entries to have at least 3
parameters, it is simpyl our assumption about what entries can appear in
the bootstrap methods table

Commit:
Added extra information to the warnings to differentaite the different cases

thk123 commented:
I've done as suggested and replaced the nested ifs with `if(!valid) continue;` style code as well as pulling out the error producing code into a separate function to reduce duplication. I've also tried to document each of the errors, so @svorenova could you see if they make sense? Will consider the tests next.

Commit:
!fixup Removed impossible condition (9fc5bfdc)

majakusber commented:
@thk123 Yes, looks very good, thank you! Will wait for the test to approve.

Commit:
Added to readme that the file is compiled using the eclipse compiler

Commit:
Use function that actually contains the lambdas

thk123 commented:
@svorenova Adjusted the tests

majakusber commented on regression/cbmc-java/lambda2/StaticMethodRef.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.function.BiFunction;
+
+public class StaticMethodRef{
+ public Integer test2(){return 2;}
```
This can be removed now it seems.

majakusber reviewed (approved):
Looks good, just one line in the test to remove but non-blocking.

Commit:
Removing redundant method from test

thk123 reviewed (approved):
I've applied my comments, so I'm happy, probably good to get a second review from a different "senior" e.g. @allredj ?

allredj commented:
@thk123 OK I'll take a look.

allredj commented on regression/cbmc-java/lambda2/StaticMethodRef.java:
```diff
@@ -0,0 +1,24 @@
+   if(ctr.equals(func2.apply("8")))
+       return func2.apply("8");
+
+   if(ctr.equals(func3.apply("0111", 2)))
```
Nit: Java format is not uniform across the file.

allredj commented on regression/cbmc-java/lambda2/StaticMethodRef.java:
```diff
@@ -0,0 +1,24 @@
+   // Uses  Integer.valueOf(String)
+   Function<String, Integer> func2  = Integer::valueOf;
+
+   BiFunction<String, Integer,  Integer> func3  = Integer::valueOf;
```
Nit: extra spaces in the generic signature.

allredj commented on regression/cbmc-java/lambda2/StaticMethodRef.java:
```diff
@@ -0,0 +1,24 @@
+       return func3.apply("0111", 2);
+
+   return ctr;
+ }
```
Bad indent

allredj commented on src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1820,122 +1827,137 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
+                 "wrong type"
+              << eom;
+      store_unknown_method_handle(
+        parsed_class, bootstrap_method_index, std::move(u2_values));
```
Question for my understanding: Why do we want to keep these unknown handles instead of just skipping the?

thk123 replied:
I think because the `invokedynamic` refers to an index in the bootstrap table which would be invalid if we didn't populate all the entries .

allredj commented on src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1820,122 +1827,137 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
+
+/// Creates an unknown method handle and puts it into the parsed_class
+/// \param parsed_class: The class whose bootstrap method handles we are using
+/// \param bootstrap_method_index: The current index in the boostrap entry table
```
`boostrap` -> `bootstrap`

allredj commented on src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1820,122 +1827,137 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
+            << "\n  method type is "
+            << id2string(pool_entry(method_type_argument.ref1).s)
+            << eom;
+    parsed_class.lambda_method_handle_map[{parsed_class.name, bootstrap_method_index}] =
```
Nit: Don't we have an 80-char limit?

allredj replied:
The way we access this map looks a bit unnatural. Why does name of `parsed_class` have to be an element of the map entry, that itself belongs to the same class. Obviously the name will always be identical for a given `parsed_class` or am I missing something? Could we have a wrapper method in `java_bytecode_parse_treet::classt` that cares about adding an element to the map instead of accessing the map directly?

thk123 replied:
Yes I agree  tagging @mgudemann to specifically review what I've done here but I agree this info is redundant and a method will make updating this in the future simpler - will do.

allredj commented on src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1820,122 +1827,137 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
+    if(
+      !(interface_type_argument.tag == CONSTANT_MethodType &&
+        method_handle_argument.tag == CONSTANT_MethodHandle &&
+        method_type_argument.tag == CONSTANT_MethodType))
```
Nit: I find conditions of the form:
```
interface_type_argument.tag != CONSTANT_MethodType ||
method_handle_argument.tag != CONSTANT_MethodHandle ||
method_type_argument.tag != CONSTANT_MethodType
```
easier to read

Commit:
Adjusted test to pass

We expect verification failure as currently we don't support lambdas so
there are lots of NPEs in the test code

allredj reviewed (approved):
Looks fine as far as I can tell. I have a couple of questions.

allredj commented on regression/cbmc-java/lambda2/test_no_crash_ecj.desc:
```diff
@@ -1,7 +1,9 @@
-^EXIT=0
+VERIFICATION FAILED
+EXIT=10
+SIGNAL=0
```
That's a bit confusing for whoever will see this test fails once lambdas start working. Could we add a comment saying what specifically we are testing here?

Commit:
Correcting formating on java file

Commit:
Adjust the interface of the bootstrap methods map

Added a todo to make private (needs adjustments to the test)

Commit:
Adjusted the if check to aid readability

Commit:
Added explanation to the test

Commit:
Fixing linter error

thk123 commented:
@allredj Addressed all comments, though have spun out as a separate ticket to make the bootstrap methods map itself private as a bit more involved

allredj commented:
@thk123 thanks!
